ASTF traffic duration with nc option (#680)

jsmoon opened the pull request:
Hi, this change is for the implementation of #680.

When `--nc` option is given, the new `m_traffic_duration` in `CGenNodeTXFIF` is used to check the `traffic duration` from `CPerProfileCtx`.
The connection established flow will be closed by `--nc` when the traffic duration is over the given `duration` value.

@hhaim, please review this change and give your feedback.

jsmoon commented:
@hhaim this update keeps the backward compatible. When `-d` and `--nc` are used, it would do the same behavior as before.
When new options `e_duration` and `t_duration` are used, they would operate independently.

| `-d` | `--nc` | `--e_duration` | `--t_duration` | description
| --- | --- | --- | --- | ---
| any | any | - | - | the same behavior as before
| - | - | 5 | - | if there is no establishment in 5 seconds, the profile will stop immediately.
| - | - | - | 10 | if traffic duration is over 10 seconds, the profile will stop immediately. -- if the first establishment time is 10 seconds, the profile stops after 10 + 10 seconds from the profile start. -- if there is no establishment forever, the profile won't stop forever. Only the client can stop the profile by request.
| - | - | 5 | 10 | if no establishment in 5 seconds, the profile will stop immediately. this profile will stop within maximum 5 + 10 seconds.
| 5 | - | - | 10 | after 5 seconds, the profile will stop if there is no active flow. -- if the flows have a long lifetime, the profile will stop after establishment time + 10 seconds.
| 5 | true | - | 10 | after 5 seconds, the profile will stop immediately. (`--t_duration` has no effect in this case)
| 20 | true | - | 10 | after establishment time + 10 seconds, the profile will stop immediately. -- if establishment time is over 10 seconds, the profile will stop immediately after 20 seconds. (maximum 20 seconds)
| 20 | true | 5 | 10 | after establishment time + 10 seconds, the profile will stop immediately when the establishment time is less than 5 seconds. (maximum 15 seconds, `-d` and `--nc` have no effect in this case)

Please review the code and give your feedback.

hhaim commented:
@jsmoon very nice table. I think there is a confusion

```
  e_duration <-> traffic_duration <-> t_duration 

e_duration - maximum time to waits for one flow to be established
traffic_duration - generate new flows
t_duration - maximum time to waits for all the flow to terminate gracefully 

```

t_duration always start **after** ```e_duration``` + ```traffic_duration```. 
This is the **maximum** duration to waits for all the flows to tear down. it waits for ```active-flows == 0```

In case ```e_duration``` is provided, this is the duration to wait for the first flow to be established, this is again the maximum duration, if the first flow was establish earlier, the traffic_duration will start at this point. 

so let take your last example:

20 | true | 5 | 10 | after establishment time + 10 seconds, the profile will stop immediately when the establishment time is less than 5 seconds. (maximum 15 seconds, -d and --nc have no effect in this case)
-- | -- | -- | -- | --


==> this is what should happen instead 

After establishment time + 20 seconds (traffic_duration which generate new flows), the profile will get into tear-down state which start a timer of 10 sec and waits for all the flows to teardown. if after 10 sec there are still ```active-flow>0 ``` the traffic will stop immediately and clean those flows. Maximum theoretical duration is 5+20+10 =35 sec

jsmoon commented:
@hhaim I agree. there is confusion. Especially, I used `t_duration` for the same meaning of `traffic duration`.
And I have been used the following terminologies as the following meaning.
- duration: the time to generate new flows from the profile begin.
- traffic duration: the time which is reported as `m_traffic_duration` in the statistics.
- first establishment time: the time that the first connection is established in the flow. it is the base time to calculate the `m_traffic_duration`,

Since the `first establishment time` is not predictable, `e_duration` should be treated as timeout value as you mentioned.

Actually, my users want to get `m_traffic_duration` for their model's traffic time. The `duration` is not appropriate to it. There is a way needed to specify the `traffic duration`.

> After establishment time + 20 seconds (traffic_duration which generate new flows), the profile will get into tear-down state which start a timer of 10 sec and waits for all the flows to teardown. if after 10 sec there are still active-flow>0 the traffic will stop immediately and clean those flows. Maximum theoretical duration is 5+20+10 =35 sec

1. the duration (as you said, traffic_duration) which generates new flows implies the `first establishment time`. So, you cannot add it to the maximum duration. i.e. it could be 30 sec.
2. if `t_duration` is used to wait until active flows are closed, I couldn't get the appropriate value for it. It's because the first establishment time is not predictable. I need a **fixed value** to guarantee the `traffic duration`. In my point of view, it's not useful that is waiting for active flows to close. (after flow generation has done)

jsmoon commented:
@hhaim what about this? based on my understanding of your suggestion, I added my solution for my requirement (red color).

![image](https://user-images.githubusercontent.com/4134762/115668133-ad633200-a381-11eb-95b9-c997ee6fb602.png)

hhaim commented:
@jsmoon  yes, I missed the statistics and the traffic generation duration (```duration```)

>> Since the first establishment time is not predictable, e_duration should be treated as timeout value as you mentioned.

agreed. so in my example maximum theoretical duration is ```20+10 =30 sec```

>> the duration (as you said, traffic_duration) which generates new flows implies the first establishment time. So, you cannot add it to the maximum duration. i.e. it could be 30 sec.

agreed. 

>> if t_duration is used to wait until active flows are closed, I couldn't get the appropriate value for it. It's because the first establishment time is not predictable. I need a fixed value to guarantee the traffic duration. In my point of view, it's not useful that is waiting for active flows to close. (after flow generation has done)

```t_duration``` is the time-out after your stop generating new flow. This is the reason I suggest to give this flexibility to the user. Today ``--nc``` set this value to zero and ```normal``` to infinite time. The user should control the timeout duration. This has very high variation depending on the setup and the scenario.

hhaim commented:
![image](https://user-images.githubusercontent.com/13034429/115688681-08862c00-a364-11eb-9617-ec09e0dfe68e.png)
@jsmoon 
I don't know why I can't see this in github, so I copy it from the email. 

Looks good. I think ```m_traffic_duration``` in case of ```--nc``` could be lower than ```-d``` as it stops immediately after the ```duration``` stoped and does not add the ```first_flow_establish_time``` 

```t_duration``` is a flexibility to limit the duration you waits for all the flow to terminate

jsmoon commented:
@hhaim I'm sorry for that. I removed it because I have realized my requirement can be accomplished without `-d` and `--nc` options.

![image](https://user-images.githubusercontent.com/4134762/115697055-2e302700-a39e-11eb-88fb-6de867639f8f.png)

I think I can get my solution by the combination of `--e_duration` and `--t_duration`.
With `-d` and `--nc`, it is changed like this.

![image](https://user-images.githubusercontent.com/4134762/115697502-94b54500-a39e-11eb-8f73-c59a11259f38.png)

I updated the table also based on above concept.

| `-d` | `--nc` | `--e_duration` | `--t_duration` | description
| --- | --- | --- | --- | ---
| any | any | - | - | the same behavior as before
| - | - | 5 | - | if there is no establishment in 5 seconds, the profile will stop immediately.
| - | - | - | 10 | if traffic duration is over 10 seconds, the profile will stop immediately. -- if the first establishment time is 10 seconds, the profile stops after 10 + 10 seconds from the profile start. -- if there is no establishment forever, the profile won't stop forever. Only the client can stop the profile by request.
| - | - | 5 | 10 | if no establishment in 5 seconds, the profile will stop immediately. otherwise, this profile will stop within maximum 5 + 10 seconds.
| 20 | - | - | 10 | after 20 seconds, the profile will stop new flow generation. and the profile will stop immediately after additional 10 seconds. (maximum 20 + 10 seconds)
| 20 | true | - | 10 | after 20 seconds, the profile will stop immediately. (`--t_duration` has no effect in this case)
| 20 | - | 5 | 10 | if no establishment in 5 seconds, the profile will stop immediately. -- after 20 seconds, the profile will stop new flow generation. and the profile will stop immediately after additional 10 seconds. (maximum 20 + 10 seconds)

jsmoon commented:
@hhaim according to your comment, I think you may prefer the following change.

`-d` | `--nc` | `--e_duration` | `--t_duration | description
-- | -- | -- | -- | --
20 | - | - | 10 | after 20 seconds, the profile will stop new flow generation. and the profile will stop immediately after additional 10 seconds. (maximum 20 + 10 seconds)
20 | true | - | 10 | after 20 seconds, the profile will stop immediately. (--t_duration has no effect in this case)
20 | - | 5 | 10 | if no establishment in 5 seconds, the profile will stop immediately. -- after 20 seconds, the profile will stop new flow generation. and the profile will stop immediately after additional 10 seconds. (maximum 20 + 10 seconds)

==>

`-d` | `--nc` | `--e_duration` | `--t_duration | description
-- | -- | -- | -- | --
20 | **any** | - | 10 | after 20 seconds, the profile will stop new flow generation. and the profile will stop immediately after additional 10 seconds. (maximum 20 + 10 seconds)
20 | **any** | 5 | 10 | if no establishment in 5 seconds, the profile will stop immediately. -- after 20 seconds, the profile will stop new flow generation. and the profile will stop immediately after additional 10 seconds. (maximum 20 + 10 seconds)

I'll prepare the PR update based on your response.

hhaim commented:
@jsmoon the chart is clear, but not the table. 

* ```--t_duration ``` duration is explicit and should take effect in all cases.  Without providing it the ```--nc``` should set the ```-t_duration``` to zero and  normal mode to infinity (backward compatible) 

* What is the meaning in the table to a row without a number for  ```-d``` ? there is always a duration 

let's take this example 
 >>20 | - | 5 | 10 | if no establishment in 5 seconds, the profile will stop immediately. -- after 20 seconds, the profile will stop new flow generation. and the profile will stop immediately after additional 10 seconds. (maximum 20 + 10 seconds)
-- | -- | -- | -- | --

it will stop after 10 sec if **not** all the flows will terminate but it could be a lower value if all flows terminated

jsmoon commented:
> What is the meaning in the table to a row without a number for -d ? there is always a duration

@hhaim it means `-d` is not specified. i.e. it is treated as `infinite`.

hhaim commented:
@jsmoon agreed.

hhaim commented:
rebased
